native_client/modelstate.cc

```diff
@@ -80,3 +80,105 @@ ModelState::decode_metadata(DecoderState* state)
   metadata->items = items.release();
   return metadata.release();
 }
+
+
+
+
+// .. we have to merge this back in, somehow..
+
+// size_t arg_max(const float *logits, size_t size) {
+//   float max = 0.0;
+//   size_t arg_max = 0;
+//   for (size_t i = 0; i < size; i++)  {
+//     if (max < logits[i]) {
+//       max = logits[i];
+//       arg_max = i;
+//     }
+//   }
+//   return arg_max;  
+// }
+
+// float entropy (const float *logits, size_t size) {
+//   float sum = 0.0f;
+//   for (size_t i = 0; i < size; i++) {
+//     if (logits[i]>0.0f)
+//       sum += logits[i] * log2(logits[i]);   
+//   }
+//   return -sum;
+// }
+
+// // what we did before
+// Metadata*
+// ModelState::decode_metadata(const vector<float>& logits)
+// {
+//   const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+//   vector<Output> out = decode_raw(logits);
+//   Output best = out[0]; 
+
+//   std::unique_ptr<Metadata> metadata(new Metadata());
+//   metadata->num_items = best.tokens.size();
+//   metadata->probability = best.probability;
+
+//   // Loop through each character, assign metadata for each
+//   std::unique_ptr<MetadataItem[]> items(new MetadataItem[metadata->num_items]());
+//   for (int i = 0; i < best.tokens.size(); ++i) {
+    
+//     // best guess from acoustic model, for the timestep corresponding to chosen character
+//     int argmax = arg_max(&logits[best.timesteps[i] * num_classes], num_classes);  
+   
+//     items[i].character = strdup(alphabet->StringFromLabel(best.tokens[i]).c_str());
+//     items[i].timestep = best.timesteps[i];
+//     items[i].start_time = best.timesteps[i] * ((float)audio_win_step / sample_rate);
+//     items[i].acoustic_char = strdup(alphabet->StringFromLabel(argmax).c_str());
+//     items[i].probability = logits[best.timesteps[i] * num_classes + best.tokens[i]];
+//     items[i].entropy = entropy(&logits[best.timesteps[i] * num_classes], num_classes);
+   
+//     if (items[i].start_time < 0) {
+//       items[i].start_time = 0;
+//     }
+//   }
+
+//   metadata->items = items.release();
+//   return metadata.release();
+// }
+
+
+
+// // partially merged
+// Metadata*
+// ModelState::decode_metadata(DecoderState* state)
+// {
+//   const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+//   vector<Output> out = decode_raw(state);
+//   Output best = out[0]; 
+
+//   std::unique_ptr<Metadata> metadata(new Metadata());
+//   metadata->num_items = best.tokens.size();
+//   metadata->probability = best.probability;
+
+//   std::unique_ptr<MetadataItem[]> items(new MetadataItem[metadata->num_items]());
+
+//   // Loop through each character
+//   for (int i = 0; i < best.tokens.size(); ++i) {
+
+//     // best guess from acoustic model, for the timestep corresponding to chosen character
+//     int argmax = arg_max(&logits[best.timesteps[i] * num_classes], num_classes);  
```

Comment:
What makes this a bit tricky is that what they previously passed to this method <probs> could be interpreted as logits but now they pass <state> which doesn't contain probs, far as i can tell.